lib/challenge_gov/accounts.ex

```diff
@@ -598,39 +599,83 @@ defmodule ChallengeGov.Accounts do
+    case determine_status_based_on_certification(previous_status, user, originator, remote_ip) do
+      {:error, decertification_result} ->
+        case decertification_result do
+          {:ok, user} ->
+            {:ok, user}
 
-    result =
-      Ecto.Multi.new()
-      |> Ecto.Multi.update(:user, changeset)
-      |> Ecto.Multi.run(:log, fn _repo, _changes ->
-        SecurityLogs.track(%{
-          originator_id: originator.id,
-          originator_role: originator.role,
-          originator_identifier: originator.email,
-          originator_remote_ip: remote_ip,
-          target_id: user.id,
-          target_type: user.role,
-          target_identifier: user.email,
-          action: "status_change",
-          details: %{previous_status: previous_status, new_status: "active"}
-        })
-      end)
-      |> Repo.transaction()
+          {:error, changeset} ->
+            {:error, changeset}
+        end
 
-    case result do
-      {:ok, _result} ->
-        {:ok, user}
+      {:ok, "active"} ->
+        changeset =
+          user
+          |> Ecto.Changeset.change()
+          |> Ecto.Changeset.put_change(:status, "active")
+          |> maybe_update_request_renewal(user)
+
+        result =
+          Ecto.Multi.new()
+          |> Ecto.Multi.update(:user, changeset)
+          |> Ecto.Multi.run(:log, fn _repo, _changes ->
+            SecurityLogs.track(%{
+              originator_id: originator.id,
+              originator_role: originator.role,
+              originator_identifier: originator.email,
+              originator_remote_ip: remote_ip,
+              target_id: user.id,
+              target_type: user.role,
+              target_identifier: user.email,
+              action: "status_change",
+              details: %{previous_status: previous_status, new_status: "active"}
+            })
+          end)
+          |> Repo.transaction()
 
-      {:error, _type, changeset, _changes} ->
-        {:error, changeset}
+        case result do
+          {:ok, %{user: user}} ->
+            {:ok, user}
+
+          {:error, _type, changeset, _changes} ->
+            {:error, changeset}
+        end
+    end
+  end
+
+  def determine_status_based_on_certification("revoked", user, approver, approver_remote_ip) do
+    if user.role != "solver",
+      do: UserController.admin_recertify_user(user, approver, approver_remote_ip)
+
+    {:ok, "active"}
+  end
+
+  def determine_status_based_on_certification("suspended", user, _approver, _approver_remote_ip) do
+    case CertificationLogs.get_current_certification(user) do
+      {:ok, certification} ->
+        # could return empty map for a solver, check for expiration
+        if !is_nil(certification.expires_at) and
+             Timex.to_unix(certification.expires_at) < Timex.to_unix(Timex.now()) do
+          decertification_result = decertify(user)
+          {:error, decertification_result}
+        else
+          {:ok, "active"}
```

Comment:
Maybe return a single atom to have a different shape to match on?

```suggestion
          :active
```